### PR TITLE
2-pass transaction application model

### PR DESCRIPTION
**Pull request by deepthiskumar**
Prioritize simple commands over zkapp account updates by separating transaction application into 2 steps:

1. Add as many simple commands (payments, delegations, fee payments, coinbase transactions) and fee payer segments from zkapp commands as you can to the diff, applying to ledger as you do so.
2. Process the account updates in applied zkapp commands. This may mean the commands fail depending on the ledger state achieved in step 1.

With this setup, all transactions that are in the mempool will always be applicable to the chain; any zkapp transaction with conflicting account updates will fail but the transaction will be included and fee extracted.  However, this solution puts additional emphasis on the user to be careful when sending simple commands and zkapp commands to the network in parallel, since they can invalidate a zkapp command through their simple commands.

Authors: @nholland94 @psteckler @mrmr1993 @MartinMinkov @deepthiskumar 

Merged the following PRs + bug fixes
https://github.com/MinaProtocol/mina/pull/11760
https://github.com/MinaProtocol/mina/pull/12377
https://github.com/MinaProtocol/mina/pull/12106
https://github.com/MinaProtocol/mina/pull/12179
https://github.com/MinaProtocol/mina/pull/12180
https://github.com/MinaProtocol/mina/pull/12355
https://github.com/MinaProtocol/mina/pull/12371
https://github.com/MinaProtocol/mina/pull/12376
https://github.com/MinaProtocol/mina/pull/12567
https://github.com/MinaProtocol/mina/pull/12554
https://github.com/MinaProtocol/mina/pull/12549

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #0000

### CODE REVIEW

File: src/lib/mina_graphql/mina_graphql.ml [hunk 1]
```diff
@@ -600,16 +600,26 @@ module Types = struct
       ~doc:
         "Transition from a source ledger to a target ledger with some fee \
          excess and increase in supply " ~fields:(fun _ ->
-        [ field "sourceLedgerHash" ~typ:(non_null ledger_hash)
-            ~doc:"Base58Check-encoded hash of the source ledger"
+        [ field "sourceFirstPassLedgerHash" ~typ:(non_null ledger_hash)
```
Comment (psteckler) [inline src/lib/mina_graphql/mina_graphql.ml:603]: nit: "first-pass" and "second-pass"

File: src/app/cli/src/init/transaction_snark_profiler.ml [hunk 2]
```diff
@@ -512,63 +535,110 @@ let pending_coinbase_stack_target (t : Transaction.t) stack =
...
+  let first_pass_target_ledgers, second_pass_target_ledgers, applied_txns =
+    apply_transactions_and_keep_intermediate_ledgers ~txn_state_view
+      sparse_ledger0 transitions
+  in
+  let final_first_pass_ledger = List.last_exn first_pass_target_ledgers in
+  List.iter second_pass_target_ledgers ~f:(fun l ->
+      assert (
+        Ledger_hash.equal
+          (Sparse_ledger.merkle_root l)
+          (Sparse_ledger.merkle_root final_first_pass_ledger) ) ) ;
```
Comment (psteckler) [inline src/app/cli/src/init/transaction_snark_profiler.ml:581]: can the Merkle root of the final 1st-pass ledger be computed just once, before the `iter` loop?

File: src/app/zkapp_test_transaction/lib/commands.ml [hunk 3]
```diff
@@ -186,12 +204,30 @@ let generate_zkapp_txn (keypair : Signature_lib.Keypair.t) (ledger : Ledger.t)
           pending_coinbase_init_stack
     }
   in
-  let witnesses, _final_ledger =
+  let witnesses =
```
Comment (psteckler) [inline src/app/zkapp_test_transaction/lib/commands.ml:207]: this is a repeat of the code above, can that be factored into a procedure?

File: src/app/zkapp_test_transaction/lib/commands.ml [hunk 4]
```diff
@@ -318,7 +354,7 @@ let test_zkapp_with_genesis_ledger_main keyfile zkapp_keyfile config_file () =
 let create_zkapp_account ~debug ~sender ~sender_nonce ~fee ~fee_payer
     ~fee_payer_nonce ~zkapp_keyfile ~amount ~memo =
   let open Deferred.Let_syntax in
-  let%bind sender_keypair = Util.keypair_of_file sender in
+  let%bind sender_keypair = Util.keypair_of_file sender ~which:"Sender" in
   let%bind fee_payer_keypair = Util.keypair_of_file fee_payer in
```
Comment (psteckler) [inline src/app/zkapp_test_transaction/lib/commands.ml:358]: if the sender keypair has a `which`, shouldn't the fee payer keypair have that, too?

File: src/lib/blockchain_snark/blockchain_snark_state.ml [hunk 5]
```diff
@@ -73,14 +73,51 @@ let non_pc_registers_equal_var t1 t2 =
...
 
+let txn_statement_ledger_hashes_equal
+    (s1 : Transaction_snark.Statement.With_sok.Checked.t)
+    (s2 : Transaction_snark.Statement.With_sok.Checked.t) =
+  Impl.make_checked (fun () ->
+      let module F = Core_kernel.Field in
+      let ( ! ) x = Impl.run_checked x in
+      let source_eq =
+        !(non_pc_registers_equal_var
+            { s1.source with pending_coinbase_stack = () }
```
Comment (psteckler) [inline src/lib/blockchain_snark/blockchain_snark_state.ml:93]: if the `pending_coinbase` parameter of this type is `unit`, which has one possible value, there's no need to replace that field

if omitting the field update leads to an ambiguous type, better to specify the complete type than use the field update

ditto for the other occurrences of this mechanism below

File: src/lib/mina_lib/mina_lib.ml [hunk 6]
```diff
@@ -630,55 +630,51 @@ let get_snarked_ledger t state_hash_opt =
         List.fold_until ~init:(Ok ()) path
```
Comment (psteckler) [inline src/lib/mina_lib/mina_lib.ml:630]: result of this fold is `()`, can use that instead of `_`

File: src/lib/consensus/intf.ml [hunk 7]
```diff
@@ -59,21 +67,26 @@ module type Blockchain_state = sig
...
 
   val genesis_ledger_hash :
-    (_, 'frozen_ledger_hash, _, _, _) Poly.t -> 'frozen_ledger_hash
+    (_, 'frozen_ledger_hash, _, _, _, _, _, _, _) Poly.t -> 'frozen_ledger_hash
 
-  val timestamp : (_, _, _, 'time, _) Poly.t -> 'time
+  val timestamp : (_, _, _, 'time, _, _, _, _, _) Poly.t -> 'time
 
-  val body_reference : (_, _, _, _, 'body_reference) Poly.t -> 'body_reference
+  val body_reference :
```
Comment (psteckler) [inline src/lib/consensus/intf.ml:88]: nit: the type parameter above is `body_ref`, rename either this function or the type parameter so parameter and function have the same name

File: src/lib/genesis_proof/dune [hunk 8]
```diff
@@ -34,6 +34,7 @@
    block_time
    kimchi_backend.pasta
    kimchi_backend.pasta.basic
+   mina_ledger
```
Comment (psteckler) [inline src/lib/genesis_proof/dune:37]: I'm able to build locally without this change

File: src/lib/ledger_proof/ledger_proof_intf.ml [hunk 9]
```diff
@@ -16,19 +16,28 @@ module type S = sig
   end]
 
   val create :
-       statement:Transaction_snark.Statement.t
+       statement:Mina_state.Snarked_ledger_state.t
```
Comment (psteckler) [inline src/lib/ledger_proof/ledger_proof_intf.ml:19]: should `statement` have a name that matches its type?

File: src/lib/ledger_proof/ledger_proof_intf.ml [hunk 10]
```diff
@@ -16,19 +16,28 @@ module type S = sig
...
     -> sok_digest:Sok_message.Digest.t
     -> proof:Proof.t
     -> t
 
   val statement_target :
```
Comment (psteckler) [inline src/lib/ledger_proof/ledger_proof_intf.ml:24]: should this function be renamed, given its new input type?

File: src/lib/ledger_proof/ledger_proof_intf.ml [hunk 11]
```diff
@@ -16,19 +16,28 @@ module type S = sig
...
-       Transaction_snark.Statement.t
+       Mina_state.Snarked_ledger_state.t
     -> ( Frozen_ledger_hash.t
        , Pending_coinbase.Stack_versioned.t
        , Mina_state.Local_state.t )
        Mina_state.Registers.t
 
-  val statement : t -> Transaction_snark.Statement.t
+  val statement : t -> Mina_state.Snarked_ledger_state.t
```
Comment (psteckler) [inline src/lib/ledger_proof/ledger_proof_intf.ml:31]: ditto the names of these functions ...

File: src/lib/mina_ledger/ledger.mli [hunk 12]
```diff
@@ -102,117 +101,14 @@ val register_mask : t -> Mask.t -> Mask.Attached.t
...
-  [@@deriving sexp]
-
-  val supply_increase : t -> Currency.Amount.Signed.t Or_error.t
-
-  val transaction : t -> Transaction.t With_status.t
-
-  val transaction_status : t -> Transaction_status.t
-end
+include
+  Mina_transaction_logic.S with type ledger := t and type location := Location.t
```
Comment (psteckler) [inline src/lib/mina_ledger/ledger.mli:105]: nice

**Review by psteckler — APPROVED** [review]
Made some comments, otherwise good.

I reviewed only these modules:
```
graphql_schema.json
src/app/cli/src/init
src/app/zkapp_test_transaction/lib
src/lib/blockchain_snark
src/lib/block_producer
src/lib/bootstrap_controller
src/lib/cli_lib
src/lib/consensus
src/lib/genesis_proof
src/lib/ledger_proof
src/lib/mina_base
src/lib/mina_block
src/lib/mina_graphql
src/lib/mina_ledger
src/lib/mina_lib
src/lib/mina_stdlib
src/lib/mina_wire_types
src/lib/mina_wire_types/mina_base
src/lib/mina_wire_types/mina_state
src/lib/mina_wire_types/test
```

**Comment by deepthiskumar** [comment]
!ci-build-me

File: src/app/zkapp_test_transaction/lib/commands.ml (hunk 4 above)
Comment (deepthiskumar) [reply src/app/zkapp_test_transaction/lib/commands.ml:358]: Fee payer is the default. I'll make it explicit

File: src/lib/ledger_proof/ledger_proof_intf.ml (hunk 10 above)
Comment (deepthiskumar) [reply src/lib/ledger_proof/ledger_proof_intf.ml:24]: I only moved the module to a different library, it is still the statement of a ledger proof

File: src/lib/blockchain_snark/blockchain_snark_state.ml (hunk 5 above)
Comment (deepthiskumar) [reply src/lib/blockchain_snark/blockchain_snark_state.ml:93]: `pending_coinbase_stack` in `s1` and `s2` has type `Pending_coinbase.Stack.var`

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-build-me

**Review by lk86 — APPROVED** [review]
Infra ✅

**Review by mrmr1993 — APPROVED** [review]
Some nits, but lgtm

File: src/lib/transaction_witness/transaction_witness.mli [hunk 13]
```diff
@@ -9,7 +9,9 @@ module Zkapp_command_segment_witness : sig
   module Stable : sig
     module V1 : sig
       type t =
-        { global_ledger : Sparse_ledger.Stable.V2.t
+        { (* PROBLEM: The parties ledger is a dependency of the final fee payment ledger... how do we compute them at the same time? *)
```
[the remark this answers is not in the document]
Comment (deepthiskumar) [reply src/lib/transaction_witness/transaction_witness.mli:12]: I think is a very old comment. Deleting it. Witnesses are generated correctly in staged ledger

File: src/lib/blockchain_snark/blockchain_snark_state.ml [hunk 14]
```diff
@@ -223,24 +260,31 @@ let%snarkydef_ step ~(logger : Logger.t)
...
+        (Protocol_state.blockchain_state previous_state).ledger_proof_statement
+      in
+      let ledger_statement_valid =
+        Impl.make_checked (fun () ->
+            Snarked_ledger_state.(
+              valid_ledgers_at_merge_checked
+                (Statement_ledgers.of_statement previous_ledger_statement)
+                (Statement_ledgers.of_statement current_ledger_statement)) )
+      in
+      (*TODO: Any assertion about the local state and sok digest
```
[the remark this answers is not in the document]
Comment (deepthiskumar) [reply src/lib/blockchain_snark/blockchain_snark_state.ml:284]: The local states are always empty across transactions. We could check that they are empty, but i'm not sure if we need to.
First pass and second pass ledgers are part of `source` and `target` in a statement and are checked by calling `valid_ledgers_at_merge_checked`.

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-build-me

**Pull request merged**